File: backend/src/mappers/simulation.py
from src.models.simulation import Simulation
from src.schema.simulation import ResponseSimulation


def simulation_to_schema(s: Simulation) -> ResponseSimulation:
    return ResponseSimulation(
        turns=int(s.turns),
        hubs=[hub.id for hub in s.hubs],
        origin=s.origin.id,
        destination=s.destination.id,
        connections=[con.id for con in s.connections],
        drones=[drone.id for drone in s.drones]
    )

File: frontend/src/schemas/simulation.ts
import { z } from "zod"
import { ConnectionSchema } from "./connection"
import { DroneSchema } from "./drone"
import { HubSchema } from "./hub"

// What the backend actually returns (arrays of UUIDs)
export const RawSimulationSchema = z.object({
	turn: z.number(),
	hubs: z.array(z.uuidv4()),
	origin: z.uuidv4(),
	destination: z.uuidv4(),
	connections: z.array(z.uuidv4()),
	drones: z.array(z.uuidv4()),
})

// Enriched simulation with full objects (built after fetching each item)
export const SimulationSchema = z.object({
	turn: z.number(),
	hubs: z.record(z.string(), HubSchema),
	origin: z.uuidv4(),
	destination: z.uuidv4(),
	connections: z.record(z.string(), ConnectionSchema),
	drones: z.record(z.string(), DroneSchema),
})
